File: backend/app/schemas/goal.py
# backend/app/schemas/goal.py

import uuid
# backend/app/schemas/goal.py

import uuid
from datetime import datetime
from typing import List, Optional
from enum import Enum
from pydantic import BaseModel, Field, validator, computed_field

from .task import TaskRead, TaskCreate # Import from the new task schema file

# --- Goal Schemas ---

class GoalBase(BaseModel):
    name: str
    category: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None

class GoalCreate(GoalBase):
    is_habit: bool = False
    tasks: List[TaskCreate] = []

class GoalUpdate(GoalBase):
    pass

class GoalRead(GoalBase):
    id: uuid.UUID
    user_id: uuid.UUID
    is_habit: bool
    tasks: List[TaskRead] = []
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def total(self) -> int:
        return len(self.tasks)

    @computed_field
    @property
    def progress(self) -> int:
        return sum(1 for task in self.tasks if task.status == "completed")

    @computed_field
    @property
    def status(self) -> str:
        if not self.tasks:
            return "On Track"
        if self.progress == self.total:
            return "Completed"
        return "On Track"

    class Config:
        from_attributes = True

# --- AI Goal Suggestion Schemas ---

class GoalType(str, Enum):
    PERSONAL = "personal"
    SHARED = "shared"

class AccountabilityType(str, Enum):
    VISUAL_PROOF = "visual_proof"
    TIME_BOUND_ACTION = "time_bound_action"
    STRICT_PHOTO_VERIFICATION = "strict_photo_verification"
    SYSTEM_VERIFIED_PUNCTUALITY = "system_verified_punctuality"

class GoalSuggestionRequest(BaseModel):
    goal_type: GoalType
    goal_name: str = Field(..., min_length=3, max_length=500)
    motivation: str = Field(..., min_length=10, max_length=1000)
    availability: List[str] = Field(..., min_items=1)
    time_commitment: str
    custom_time: Optional[str] = None
    accountability_type: AccountabilityType
    time_window: Optional[str] = None
    partner_name: Optional[str] = None
    
    @validator('availability')
    def validate_availability(cls, v):
        valid_options = [
            "Mornings (6-9 AM)", "Lunchtime (12-2 PM)", "Evenings (6-9 PM)",
            "Weekends only", "I'm flexible", "Early Mornings (6-8 AM)",
            "Lunch Break (12-2 PM)", "Weekend Mornings", "Weekend Evenings",
            "We're flexible"
        ]
        for item in v:
            if item not in valid_options:
                raise ValueError(f"Invalid availability option: {item}")
        return v

class SuggestedTask(BaseModel):
    task_name: str = Field(..., min_length=5, max_length=200)
    description: str = Field(..., min_length=10, max_length=500)
    repeat_frequency: str = Field(..., description="e.g., 'Daily', '3 times a week'")

class GoalSuggestionResponse(BaseModel):
    goal_type: str = Field(..., description="'Project' or 'Habit'")
    tasks: List[SuggestedTask] = Field(..., min_items=3, max_items=5)
    success_tips: List[str] = Field(..., min_items=2, max_items=4)
    generated_at: str
    model_version: str = "gemini-3-flash"

class OnboardingGoalPlanRequest(BaseModel):
    goal_title: str
    goal_description: str
    contextual_answers: dict

File: src/schemas/goal.ts
// src/schemas/goal.ts
import { z } from 'zod';
import {
  GoalWizardRequestSchema,
  StrategicQuestionSchema,
  QuestionsResponseSchema,
  AnswersSubmissionRequestSchema,
  DuotrakGoalPlanSchema,
  GoalPlanResponseSchema,
  TaskVerificationModeSchema,
  TaskProofGuidanceSchema,
  TaskPartnerInvolvementSchema,
  TaskCadenceSchema,
  HabitConfigSchema,
  MilestoneConfigSchema,
  TargetDateConfigSchema,
  type GoalWizardRequest,
  type StrategicQuestion,
  type QuestionsResponse,
  type AnswersSubmissionRequest,
  type DuotrakGoalPlan,
  type GoalPlanResponse,
  type TaskVerificationMode,
  type TaskProofGuidance,
  type TaskPartnerInvolvement,
  type TaskCadence,
  type HabitConfig,
  type MilestoneConfig,
  type TargetDateConfig,
} from "../../packages/contracts/src/goalCreation";

export {
  GoalWizardRequestSchema,
  StrategicQuestionSchema,
  QuestionsResponseSchema,
  AnswersSubmissionRequestSchema,
  DuotrakGoalPlanSchema,
  GoalPlanResponseSchema,
  TaskVerificationModeSchema,
  TaskProofGuidanceSchema,
  TaskPartnerInvolvementSchema,
  TaskCadenceSchema,
  HabitConfigSchema,
  MilestoneConfigSchema,
  TargetDateConfigSchema,
};

export type {
  GoalWizardRequest,
  StrategicQuestion,
  QuestionsResponse,
  AnswersSubmissionRequest,
  DuotrakGoalPlan,
  GoalPlanResponse,
  TaskVerificationMode,
  TaskProofGuidance,
  TaskPartnerInvolvement,
  TaskCadence,
  HabitConfig,
  MilestoneConfig,
  TargetDateConfig,
};

export const TaskSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable().optional(),
  status: z.string(),
  due_date: z.string().datetime().nullable(),
  goal_id: z.string().uuid(),
  repeat_frequency: z.string().nullable().optional(),
  time_window: z.string().nullable().optional(),
  accountability_type: z.string().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export const GoalBaseSchema = z.object({
  name: z.string(),
  description: z.string().nullable().optional(),
  motivation: z.string().nullable().optional(),
  category: z.string().nullable(),
  icon: z.string().nullable(),
  color: z.string().nullable(),
});

export const GoalReadSchema = GoalBaseSchema.extend({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  isHabit: z.boolean(),
  tasks: z.array(TaskSchema),
  availability: z.array(z.string()).nullable().optional(),
  timeCommitment: z.string().nullable().optional(),
  accountabilityType: z.string().nullable().optional(),
  templateSourceId: z.string().optional(),
  templateSourceSlug: z.string().optional(),
  templateSourceVersion: z.number().optional(),
  templateSourceTitle: z.string().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  // Computed fields from backend
  total: z.number().int(),
  progress: z.number().int(),
  status: z.string(),
});

export const GoalCreateSchema = GoalBaseSchema.extend({
  isHabit: z.boolean().default(false),
  goalType: z.enum(["habit", "target-date", "milestone"]).optional(),
  endDate: z.string().datetime().nullable().optional(),
  templateSourceId: z.string().optional(),
  templateSourceSlug: z.string().optional(),
  templateSourceVersion: z.number().optional(),
  templateSourceTitle: z.string().optional(),
  availability: z.array(z.string()).optional(),
  timeCommitment: z.string().optional(),
  accountabilityType: z.string().optional(),
  tasks: z.array(z.object({
    name: z.string(),
    description: z.string().optional(),
    repeatFrequency: z.string().optional(),
    timeWindow: z.string().optional(),
    accountabilityType: z.string().optional(),
    verificationMode: z.string().optional(),
    timeWindowStart: z.string().optional(),
    timeWindowEnd: z.string().optional(),
    timeWindowDurationMinutes: z.number().int().positive().optional(),
    requiresPartnerReview: z.boolean().optional(),
  })),
});

// For updates, we allow a subset of the base fields.
// Note: We are not allowing tasks to be updated via this endpoint for now.
export const GoalUpdateSchema = z.object({
  name: z.string().optional(),
  description: z.string().optional(),
  motivation: z.string().optional(),
  category: z.string().optional(),
  icon: z.string().optional(),
  color: z.string().optional(),
  goalType: z.enum(["habit", "target-date", "milestone"]).optional(),
  endDate: z.string().datetime().optional(),
  templateSourceId: z.string().optional(),
  templateSourceSlug: z.string().optional(),
  templateSourceVersion: z.number().optional(),
  templateSourceTitle: z.string().optional(),
  availability: z.array(z.string()).optional(),
  timeCommitment: z.string().optional(),
  accountabilityType: z.string().optional(),
});

export type TaskRead = z.infer<typeof TaskSchema>;
export type GoalRead = z.infer<typeof GoalReadSchema>;
export type GoalCreate = z.infer<typeof GoalCreateSchema>;
export type GoalUpdate = z.infer<typeof GoalUpdateSchema>;


// --- AI Goal Suggestion Schemas ---
export const GoalSuggestionRequestSchema = z.object({
  goalType: z.enum(["personal", "shared"]),
  goalName: z.string(),
  motivation: z.string(),
  availability: z.array(z.string()),
  timeCommitment: z.string(),
  customTime: z.string().optional(),
  accountabilityType: z.string(),
  timeWindow: z.string().optional(),
  partnerName: z.string().optional(),
});

export const SuggestedTaskSchema = z.object({
  taskName: z.string(),
  description: z.string(),
  repeatFrequency: z.string(),
});

export const GoalSuggestionResponseSchema = z.object({
  goalType: z.string(),
  tasks: z.array(SuggestedTaskSchema),
  successTips: z.array(z.string()),
  generatedAt: z.string(),
  modelVersion: z.string(),
});

export type GoalSuggestionRequest = z.infer<typeof GoalSuggestionRequestSchema>;
export type SuggestedTask = z.infer<typeof SuggestedTaskSchema>;
export type GoalSuggestionResponse = z.infer<typeof GoalSuggestionResponseSchema>;
